plutus-tx/src/PlutusTx/List.hs

```diff
@@ -49,6 +65,57 @@ map f l = case l of
     []   -> []
     x:xs -> f x : map f xs
 
+{-# INLINABLE and #-}
+-- | Returns the conjunction of a list of Bools.
+and :: [Bool] -> Bool
+and = \case
+    []     -> True
+    x : xs -> if x then and xs else False
+
+{-# INLINABLE or #-}
+-- | Returns the disjunction of a list of Bools.
+or :: [Bool] -> Bool
+or = \case
+    []     -> False
+    x : xs -> if x then True else or xs
+
+{-# INLINABLE any #-}
+-- | Determines whether any element of the structure satisfies the predicate.
+any :: (a -> Bool) -> [a] -> Bool
+any f = go
+  where
+    go = \case
+        []     -> False
+        x : xs -> if f x then True else go xs
+
+{-# INLINABLE all #-}
+-- | Determines whether all elements of the list satisfy the predicate.
+all :: (a -> Bool) -> [a] -> Bool
+all f = go
+  where
+    go = \case
+        []     -> True
+        x : xs -> if f x then go xs else False
+
+{-# INLINABLE elem #-}
+-- | Does the element occur in the list?
+elem :: Eq a => a -> [a] -> Bool
+elem = any . (==)
+
+{-# INLINABLE notElem #-}
+-- | The negation of `elem`.
+notElem :: Eq a => a -> [a] -> Bool
+notElem a = not . elem a
+
+{-# INLINABLE find #-}
+-- | Returns the leftmost element matching the predicate, or `Nothing` if there's no such element.
+find :: (a -> Bool) -> [a] -> Maybe a
+find f = go
```

Comment:
Optimize `findIndex` as well?

Reply:
> `findIndex` is already optimized

But it traverses the entire list, because `findIndices` that it calls does so?

Reply:
I don't think so. `findIndices` is implemented directly, not in terms of `foldr`.

Reply:
Ah, sorry, I had a brain fog. Yes `findIndex` is un-optimized.

Reply:
Sure, but it still traverses the entire list, it's the point of the function to return all the indices, for which it needs to traverse the whole list! `findIndex` then throws all of them but the first one away, but Plutus is strict, hence the list of indices still has to materialize first.

I really hope it's not one of these times when I'm saying utter nonsense with a straight face :smiley:

Reply:
Eek, didn't see your message, ignore my last one.